enhancements/testing/ephemeral-agent-installer-e2e-tests.md

```diff
@@ -0,0 +1,373 @@
+cluster 0, this is highly undesirable.
+
+**Implementation Plan:**
+
+* Create a new routine to run Assisted Installer outside of k8s.
+
+* Create a new routine to launch a VM with the built ephemeral agent based installer
+image.
+
+### API Extensions
+
+Not applicable at this time.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Please refer to the Design Details section.
+
+### Risks and Mitigations
+
+The direction of this proposal is now to extend and refactor parts of dev-scripts
+for the agent-based installer.
+
+** There will be now a third team using dev-scripts for slightly different purposes.
+Which means the risk of breaking things in the central framework now is larger and
+has a larger "splash zone."
+
+We can mitigate this for the new ephemeral agent installer by adding e2e tests into the
+dev-scripts repository that run the agent flows. The purpose of this would be to find 
+breakages in the tests across the different teams.
+
+** The dev-scripts repository clones the [metal-dev-env](https://github.com/metal3-io/metal3-dev-env) 
+repository for most of its Ansible roles. Which means there is a risk of changes in this 
+repository can break things in our repository as well. We have less control over this repository 
+than others inside Openshift.
+
+This is partially mitigated already because the repository is cloned at a specific
+Git SHA. To further this we can add a test in dev-scripts that runs maybe once or twice
+a week that clones the HEAD of this repository. We would be alerted for breakages more
+frequently and we can mitigate them on a case-by-case basis.
+
+
+## Design Details
+
+### Identified Overlap Points for Agent Installer
+
+During the discussions with the Metal Platform team it was identified that the following
+components of dev-scripts are the likely points of overlap for the new agent installer project:
+
+* The [01_install_requirements.sh](https://github.com/openshift-metal3/dev-scripts/blob/master/01_install_requirements.sh) 
+and [02_configure_host.sh](https://github.com/openshift-metal3/dev-scripts/blob/master/02_configure_host.sh) 
+scripts that are run using `make`.
+* Ansible playbooks to set up VMs such as: [vm_setup_vars.yml](https://github.com/openshift-metal3/dev-scripts/blob/master/vm_setup_vars.yml)
+
+These will be the things that we will focus on first when refactoring parts of dev-scripts.
+
+### Overall Components of the e2e Framework
+
+Using the general flow of an e2e test we can sort the overall components of the e2e framework.
+A general flow of an e2e test has the following phases:
+
+* Pre Environment Setup: 
+** Setting up VMs, networks, locations of images, image registries, etc
+* Build: 
+** Building the agent-based installer ISO and container images associated with it
+* Test:
+**  Test scenarios with pass/failure outputs
+* Reporting:
+** Reports comprised of test results
+* Environment Teardown
+** Removing VMs, networks, etc to start fresh again
+
```

Comment:
The parts that are likely to be reused in the framework would be pre-environment setup and teardown.  I would focus most of the time here to make them easy to customize and reuse. Each project would tend to having specific requirements for build, test, and reporting. Having a generic way to be able to plug those in would be a big help.

Reply:
I took reporting out of the scope of this enchancement. In the future linking into the existing reports should be looked at.